File: src/node/halchemy/src/lib/follower.ts
import {Api} from "./api";
import {HalResource, HalLink} from "hal-types";
import {Requester} from "./requester";

export class Follower {

    private readonly _api: Api
    private _resource: HalResource

    constructor(api: Api, resource: HalResource) {
        this._api = api
        this._resource = resource
    }

    to(rel: string): Requester {
        if (!(rel in this._resource._links)) {
            throw new Error(`This resource does not have a link relation named '${rel}'`)
        }
        return new Requester(this._api, {resource: this._resource, rel: rel})
    }
}

File: src/python/halchemy/lib/follower.py
"""
This module provides the Follower class, which is used to follow links from a HalResource.
"""

from .resource import HalResource
from .requester import Requester


class Follower:
    def __init__(self, api, resource: HalResource):
        self.api = api
        self.resource = resource

    def to(self, rel: str) -> Requester:
        if rel not in self.resource['_links']:
            raise KeyError(f"This resource does not have a link relation named '{rel}'")
        return Requester(self.api, (self.resource, rel))
